File: src/sp/it/pl/util/reactive/SetƑ1.java
package sp.it.pl.util.reactive;

import java.util.HashSet;
import java.util.function.Consumer;
import org.reactfx.Subscription;
import sp.it.pl.util.functional.Functors.Ƒ1;

/** Set of consumers/functions taking 1 parameter. For use as a collection of handlers/listeners. */
public class SetƑ1<I> extends HashSet<Consumer<I>> implements Ƒ1<I,Void> {

	public SetƑ1() {
		super(2);
	}

	@Override
	public Void apply(I input) {
		forEach(c -> c.accept(input));
		return null;
	}

	public Subscription addS(Consumer<I> r) {
		add(r);
		return () -> remove(r);
	}

}

File: src/sp/it/pl/util/reactive/SetƑ1.kt
package sp.it.pl.util.reactive

import org.reactfx.Subscription
import sp.it.pl.util.functional.Functors.Ƒ1
import sp.it.pl.util.functional.invoke
import java.util.HashSet
import java.util.function.Consumer

/** Set of actions taking 1 parameter. Use as a collection of handlers/listeners. */
class SetƑ1<I>: HashSet<Consumer<I>>(2), Ƒ1<I, Unit> {

    override fun apply(input: I) {
        forEach { it(input) }
    }

    fun addS(r: Consumer<I>): Subscription {
        add(r)
        return Subscription { remove(r) }
    }

}